### PR TITLE
Create base collection layout and assets

### PR DESCRIPTION
**Pull request by ehelms**
Borrowed and minor adaptations from https://github.com/theforeman/foreman-ansible-modules to match as close as possible to test/release/management of our collections.

### CODE REVIEW

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+```
+
+##### KATELLO/FOREMAN VERSION
+<!--- Paste, BELOW THIS COMMENT, verbatim output from "rpm -q tfm-rubygem-katello foreman"-->
+```
+
+```
+
+##### APYPIE VERSION
```
Comment (mdellweg) [inline .github/ISSUE_TEMPLATE/bug_report.md]: If i understand correctly, this should be the version of FAM here.

File: .github/workflows/main.yml [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    strategy:
+      fail-fast: false
+      matrix:
+        python: ["2.7", "3.5", "3.6", "3.7"]
+        ansible:
+          - stable-2.7
```
Comment (mdellweg) [inline .github/workflows/main.yml]: This is going to be a collection to begin with, right?
Then you should not test ansible 2.7
However 2.10 would be interesting. ;)

File: .github/workflows/main.yml [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+      - name: Install Ansible devel specific dependencies
+        run: ansible-galaxy collection install community.general
+        if: matrix.ansible == 'devel'
+      - name: Run basic tests
+        run: make test-crud test-other
+      - name: Run Ansible devel specific tests
+        run: make SANITY_OPTS="--local" PYTHON_VERSION=${{ matrix.python }} sanity dist-test
+        if: matrix.ansible == 'devel'
+
+  checkmode:
```
Comment (mdellweg) [inline .github/workflows/main.yml]: Not sure this is a thing if we don't have modules here.

File: Makefile [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	rm -rf build
+
+doc-setup:
+	pip install -r docs/requirements.txt
+doc:
+	make -C docs html
+
+FORCE:
+
+.PHONY: help dist lint sanity test test-crud test-check-mode test-other setup test-setup FORCE
```
Comment (mdellweg) [inline Makefile]: This reminds me that i wanted to look into reusable makefile includes.

File: docs/developing.md [hunk 5]
```diff
@@ -0,0 +1,101 @@
+# How to write modules
```
Comment (mdellweg) [inline docs/developing.md]: This whole file seems very specific to FAM.

File: docs/index.rst [hunk 6]
```diff
@@ -0,0 +1,28 @@
+.. Foreman Ansible Modules documentation master file, created by
+   sphinx-quickstart on Tue Dec 11 10:00:39 2018.
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to Foreman Ansible Modules's documentation!
```
Comment (mdellweg) [inline docs/index.rst]: ... Operations ...

File: docs/releasing.md [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+* Create a pull request to update `galaxy.yml` with:
+  * Version bump
+  * Updated list of authors.
+
+    You can generate that list via `git shortlog -e -s | sed -E 's/^\s+[[:digit:]]+\s+(.*)$/  - "\1"/'`.
+
+* After merging, tag the merge commit with `v<version number>` (i.e. v0.1.1).
+  Please use signed tags.
+
+Consider filing a packaging PR on [foreman-packaging](https://github.com/theforeman/foreman-packaging).
```
Comment (mdellweg) [inline docs/releasing.md:18]: :D This file seems generic enough.

File: docs/testing.md [hunk 8]
```diff
@@ -0,0 +1,103 @@
+# Testing Foreman Ansible Modules
```
Comment (mdellweg) [inline docs/testing.md]: Again, (in parts) very FAM specific

**Review by mdellweg — COMMENTED** [review]
This is by far not a complete review. i.e. the whole docs processing is a black box to me.
However, while all looks quite familiar, can it be, that you copied the files from a rather old instantiation of FAM?

**Comment by ehelms** [comment]
@mdellweg bah you are right, I was on the deprecated master branch when I copied (@evgeni can we delete the master branch from FAM to avoid confusion?) and I will update this based on latest develop

**Comment by evgeni** [comment]
@ehelms I did long ago: https://github.com/theforeman/foreman-ansible-modules/branches (and closed some PRs doing so **cough**)

**Comment by ehelms** [comment]
@evgeni wth, figured my script would have errored trying to fetch it, this is why you shouldn't let me play with Ansible code repos :)

**Comment by evgeni** [comment]
Well, if your local checkout still had a `master`, it might have taken that? Or was it a fresh clone?

File: .github/workflows/main.yml (hunk 3 above)
Comment (ehelms) [reply .github/workflows/main.yml]: I think we will in the end, so I debated whether to leave it for when that time comes or just be ready for it.

**Comment by ehelms** [comment]
Latest update, I have pushed empty github workflows to ensure what gets merged here works in some fashion from the outset. I have updated the Makefile to account for this and adjusted the test templates as well.

File: .github/ISSUE_TEMPLATE/bug_report.md [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+##### ISSUE TYPE
+ - Bug Report
+
+##### ANSIBLE VERSION
+<!-- Paste, BELOW THIS COMMENT, verbatim output from "ansible --version"-->
+```
+
+```
+
+##### COLLECTION VERSION
```
Comment (mdellweg) [inline .github/ISSUE_TEMPLATE/bug_report.md:20]: :+1:

File: .github/workflows/main.yml [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+    strategy:
+      fail-fast: false
+      matrix:
+        python:
+          - "2.7"
+          - "3.6"
+          - "3.7"
+          - "3.8"
+        ansible:
+          - stable-2.8
```
Comment (mdellweg) [inline .github/workflows/main.yml]: Are you sure, you want to start a new collection with ansible 2.8 in mind?

File: Dockerfile [hunk 11]
```diff
@@ -0,0 +1,8 @@
+FROM ansible/ansible-runner
+
+RUN echo "tsflags=nodocs" >> /etc/yum.conf && \
+		yum -y install git
+
+RUN git clone https://github.com/theforeman/foreman-operations-collection ~/.ansible/plugins/modules/
+
+RUN cd ~/.ansible/plugins/modules && pip install -r requirements-dev.txt
```
Comment (mdellweg) [inline Dockerfile]: Are we using this container anywhere?

File: README.md [hunk 12]
```diff
@@ -1,2 +1,50 @@
...
+
+### Supported Ansible Versions
+
+The supported Ansible versions are aligned with currently maintained Ansible versions. You can find the list of maintained Ansible versions [here](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#release-status). Older Ansible versions (Ansible >= 2.3) should work but are not tested.
+
+As we're using Ansible's [documentation fragment](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-fragments) feature, that was introduced in Ansible 2.8, `ansible-doc` prior to 2.8 won't be able to display the module documentation, but the modules will still run fine with `ansible` and `ansible-playbook`.
+
+### Supported Python Versions
+
+Starting with Ansible 2.7, Ansible only supports Python 2.7 and 3.5 (and higher). These are also the only Python versions we develop and test the modules against.
```
Comment (mdellweg) [inline README.md]: 2.7 (Ansible) is definitely not supported anymore...

File: README.md [hunk 13]
```diff
@@ -1,2 +1,50 @@
...
+### Supported Foreman and plugins versions
+
+Ansible assets should support any currently stable Foreman release and the matching set of plugins.
+Some modules have additional features/arguments that are only applied when the corresponding plugin is installed.
+
+We actively test the modules against the latest stable Foreman release and the matching set of plugins.
+
+### Supported Ansible Versions
+
+The supported Ansible versions are aligned with currently maintained Ansible versions. You can find the list of maintained Ansible versions [here](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#release-status). Older Ansible versions (Ansible >= 2.3) should work but are not tested.
```
Comment (mdellweg) [inline README.md]: 2.3 ???
This is either >= 2.8 or >=2.9, because collections.

File: README.md [hunk 14]
```diff
@@ -1,2 +1,50 @@
...
+
+### Supported Python Versions
+
+Starting with Ansible 2.7, Ansible only supports Python 2.7 and 3.5 (and higher). These are also the only Python versions we develop and test the modules against.
+
+### Known issues
+
+## Installation
+
+There are currently two ways to use the modules in your setup: install from Ansible Galaxy or via RPM.
```
Comment (mdellweg) [inline README.md]: I can imagine RPM is planned...

File: README.md [hunk 15]
```diff
@@ -1,2 +1,50 @@
...
+
+### Installation from Ansible Galaxy
+
+You can install the collection from [Ansible Galaxy](https://galaxy.ansible.com/theforeman/foreman) by running `mazer install theforeman.foreman` (Ansible 2.8) or `ansible-galaxy collection install theforeman.operations` (Ansible 2.9 and later).
+
+After the installation, the assets are available as `theforeman.operations.<name>`. Please see the [Using Ansible collections documentation](https://docs.ansible.com/ansible/devel/user_guide/collections_using.html) for further details.
+
+### Installation via RPM
```
Comment (mdellweg) [inline README.md:38]: In this section at least the numbers are wrong.

**Review by mdellweg — COMMENTED** [review]
Found some more inconsistencies...

File: README.md (hunk 12 above)
Comment (ehelms) [reply README.md]: Agreed, I do not think this is intended to say that. I think is this just giving history that in Ansible 2.7 they started only supporting Python 2.7 and 3.5 (or higher). That is, putting the blame on Ansible. I can just reword this :)

File: README.md (hunk 14 above)
Comment (ehelms) [reply README.md]: Correct.

Comment (ehelms) [reply README.md]: But I can rip all of that out and until I get that far if y'all prefer.

File: .github/workflows/main.yml (hunk 10 above)
Comment (ehelms) [reply .github/workflows/main.yml]: Dropped 2.8. Starting clean.

**Comment by mdellweg** [comment]
If not supporting 2.8, you can rip out all the mazer hacks.

**Comment by ehelms** [comment]
> If not supporting 2.8, you can rip out all the mazer hacks.

Great point, deleted them.

File: requirements-common.txt [hunk 16]
```diff
@@ -0,0 +1,2 @@
+ansible>=2.3
```
Comment (mdellweg) [inline requirements-common.txt]:
```suggestion
ansible>=2.9
```

**Review by mdellweg — APPROVED** [review]
I cannot speak so much about the doc generation process. @evgeni ?

But in general, i think this is something we can have a good start from.

File: docs/_gh_include/header.inc [hunk 17]
```diff
@@ -0,0 +1,17 @@
...
+  <title>Foreman Ansible Modules documentation index</title>
+  <meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
+  <link rel="stylesheet" media="all" href="https://theforeman.org/static/css/bootstrap.min.css"/>
+  <link rel="stylesheet" media="all" href="https://theforeman.org/static/css/bootstrap-responsive.min.css"/>
+  <link rel="stylesheet" media="all" href="https://theforeman.org/static/css/style.css"/>
+  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/font-awesome/4.3.0/css/font-awesome.min.css">
+  </head>
+
+  <body style="margin:0px 15px">
+    <h1>Foreman Ansible Modules</h1>
```
Comment (evgeni) [inline docs/_gh_include/header.inc]: wrong name in that file ;)

File: docs/_templates/plugin.rst.j2 [hunk 18]
```diff
@@ -0,0 +1,469 @@
+:source: @{ source }@
```
Comment (evgeni) [inline docs/_templates/plugin.rst.j2]: that file should not be there anymore

File: docs/_themes/ansible_rtd_theme/static/ansible_rtd.css [hunk 19]
```diff
@@ -0,0 +1,515 @@
+@import url('css/theme.css');
```
Comment (evgeni) [inline docs/_themes/ansible_rtd_theme/static/ansible_rtd.css]: that folder (`_themes`) should not be there anymore

File: docs/index.rst [hunk 20]
```diff
@@ -0,0 +1,32 @@
+.. Foreman Operations Collection documentation main file, created by
+   sphinx-quickstart on Tue Dec 11 10:00:39 2018.
+   You can adapt this file completely to your liking, but it should at least
+   contain the root `toctree` directive.
+
+Welcome to Foreman Operations Collection' documentation!
```
Comment (evgeni) [inline docs/index.rst]:
```suggestion
Welcome to Foreman Operations Collection documentation!
```

File: docs/index.rst [hunk 21]
```diff
@@ -0,0 +1,32 @@
...
+   plugins/index
+   Filters <filters>
+   roles/index
+
+.. toctree::
+   :maxdepth: 2
+   :caption: Developer documentation
+
+   developing
+   api
```
Comment (evgeni) [inline docs/index.rst]: I don't think we have an `api` file yet.

File: galaxy.yml [hunk 22]
```diff
@@ -0,0 +1,17 @@
+namespace: "theforeman"
+name: "operations"
+description: Ansible assets to manage Foreman operations such as install, upgrade and maintenance tasks
```
Comment (evgeni) [inline galaxy.yml]: starting a description of an ansible collection with "ansible assets" sounds kinda repetitive?

```suggestion
description: Foreman operations tasks such as install, upgrade and maintenance
```

**Comment by ehelms** [comment]
Comments addressed, I appreciate the patience reviewing a big PR to get things started.

**Comment by evgeni** [comment]
@ehelms i meant the full docs/_themes and docs/_templates folders can go, they aren't present in the current FAM code.

**Comment by ehelms** [comment]
> @ehelms i meant the full docs/_themes and docs/_templates folders can go, they aren't present in the current FAM code.

You should have just said that :) THEY GONE

**Comment by evgeni** [comment]
I thought I did!

File: .github/workflows/main.yml [hunk 23]
```diff
@@ -0,0 +1,114 @@
...
+          key: ${{ runner.os }}-pip-${{ hashFiles('**/requirements*.txt') }}
+          restore-keys: |
+            ${{ runner.os }}-pip-
+      - name: Install libyaml-dev for PyYAML
+        run: sudo apt-get install -y libyaml-dev
+      - name: Install Ansible
+        run: pip install --upgrade git+https://github.com/ansible/ansible.git@${{ matrix.ansible }}
+      - name: Install dependencies
+        run: make test-setup
+      - name: fake redhat-uep.pem for redhat_manifest module
```
Comment (evgeni) [inline .github/workflows/main.yml]: This step won't be necessary here

File: .github/workflows/main.yml [hunk 24]
```diff
@@ -0,0 +1,114 @@
...
+            ${{ runner.os }}-pip-
+      - name: Install dependencies
+        run: pip install --upgrade -r requirements-lint.txt
+      - name: Run lint
+        run: make lint
+      - name: Upload collection artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: collection
+          path: theforeman-foreman-*.tar.gz
```
Comment (evgeni) [inline .github/workflows/main.yml]: This path needs updating for the new name

File: README.md [hunk 25]
```diff
@@ -1,2 +1,50 @@
...
+Ansible assets should support any currently stable Foreman release and the matching set of plugins.
+Some modules have additional features/arguments that are only applied when the corresponding plugin is installed.
+
+We actively test the modules against the latest stable Foreman release and the matching set of plugins.
+
+### Supported Ansible Versions
+
+The supported Ansible versions are aligned with currently maintained Ansible versions. You can find the list of maintained Ansible versions [here](https://docs.ansible.com/ansible/latest/reference_appendices/release_and_maintenance.html#release-status).
+
+As we're using Ansible's [documentation fragment](https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_documenting.html#documentation-fragments) feature, that was introduced in Ansible 2.8, `ansible-doc` prior to 2.8 won't be able to display the module documentation, but the modules will still run fine with `ansible` and `ansible-playbook`.
```
Comment (evgeni) [inline README.md]: This section is irrelevant if we only support 2.9+

File: README.md [hunk 26]
```diff
@@ -1,2 +1,50 @@
...
+
+### Known issues
+
+## Installation
+
+There are currently two ways to use the modules in your setup: install from Ansible Galaxy or via RPM. RPM installation is planned for a future release.
+
+### Installation from Ansible Galaxy
+
+You can install the collection from [Ansible Galaxy](https://galaxy.ansible.com/theforeman/foreman) by running `ansible-galaxy collection install theforeman.operations`.
```
Comment (evgeni) [inline README.md]: Wrong galaxy link ;)

**Review by evgeni — CHANGES_REQUESTED** [review]

File: requirements-common.txt (hunk 16 above)
Comment (evgeni) [reply requirements-common.txt]: `requirements-common` doesn't exist in FAM anymore, and we don't define a dependecy on ansible besides in the docs.

**Comment by evgeni** [comment]
Thanks @ehelms!

I think the only two remaining points are:
* what is this `Dockerfile` used for? We don't have one in FAM.
* `requirements-common.txt` should go away as in FAM

**Comment by ehelms** [comment]
> I think the only two remaining points are:
> 
> * what is this `Dockerfile` used for? We don't have one in FAM.
> * `requirements-common.txt` should go away as in FAM

Removed

**Review by evgeni — APPROVED** [review]

**Pull request merged**